textract/layout/analysis.py

```diff
@@ -0,0 +1,240 @@
+
+		else:
+			_, src = cv2.threshold(src, 0, 255, cv2.THRESH_BINARY+cv2.THRESH_OTSU)
+
+		return src
+
+	def _calculate_bbox(self, src):
+		img = src.copy()
+		bboxes = []
+
+		# get bounding box and keep merging until no intersection
+		while True:
+			(img_contours, _) = cv2.findContours(img.copy(), cv2.RETR_EXTERNAL,cv2.CHAIN_APPROX_SIMPLE)
+			
+			# make mask 
+			mask = np.zeros(img.shape[:2], dtype=np.uint8)
+
+			for cnt in img_contours:
+				# adding bouding box 
+				x,y,w,h = cv2.boundingRect(cnt)           
+				cv2.rectangle(mask, (x+1,y+1), (x+w-1,y+h-1), (255), -1)
+
+			(mask_contours, _) = cv2.findContours(mask.copy(), cv2.RETR_EXTERNAL,cv2.CHAIN_APPROX_SIMPLE)
+
+			if len(img_contours) == len(mask_contours):
+				# no intersection
+				for i in range(len(mask_contours)):
+					x, y, w, h = cv2.boundingRect(mask_contours[i])
+					bbox = BBox(i, x, y, w, h)
+					bboxes.append(bbox)
+				break
+			else:
+				# still have intersection, keep iterating
+				img = mask.copy()
+
+		return bboxes
+
+	def _build_graph(self, bboxes):
+		'''
+		1. Line segment a comes before line segment b 
+			- if their ranges of x-coordinates overlap 
+		and - if line segment a is above line segment b on the page
+		
+		2. Line segment a comes before line segment b 
+			- if a is entirely to the left of b
+		and - there does not exist a line segment c whose y-coordinates are between a and b and whose range of xcoordinates overlaps both a and b.
+		'''
+
+		graph = Graph(len(bboxes))
+
+		# sort by (y, x) coordinate (top down -> left right)
+		bboxes_sorty = sorted(bboxes, key = lambda box : (box.get_coords()[1], box.get_coords()[0]))
+		lookup_table = {}
+
+		# rule 1
+		for i in range(len(bboxes_sorty)):
+			bbox1 = bboxes_sorty[i]
+			lookup_table[bbox1.get_id()] = i
+
+			for j in range(i+1, len(bboxes_sorty)):
+				bbox2 = bboxes_sorty[j]
+
+				if self._find_overlap(bbox1, bbox2) and not graph.check_cyclic(bbox1.get_id(), bbox2.get_id()):
+					graph.add_edge(bbox1.get_id(), bbox2.get_id())
+
+		# sort by (x, y) coordinate (left right -> top down)
+		bboxes_sortx = sorted(bboxes, key = lambda box : (box.get_coords()[0], box.get_coords()[1]))
+		
+		# rule 2
+		for i in range(len(bboxes_sortx)):
+			bbox1 = bboxes_sortx[i]
+
+			for j in range(i+1, len(bboxes_sortx)):
+				bbox2 = bboxes_sortx[j]
+
+				if not self._find_overlap(bbox1, bbox2):
+					is_neighbor = True
+					for k in range(min(lookup_table[bbox1.get_id()], lookup_table[bbox2.get_id()]), max(lookup_table[bbox1.get_id()], lookup_table[bbox2.get_id()])):
+						if self._find_overlap(bbox1, bboxes_sorty[k]) and self._find_overlap(bbox2, bboxes_sorty[k]):
+							is_neighbor = False
+							break
+
+					if is_neighbor and not graph.check_cyclic(bbox1.get_id(), bbox2.get_id()):
+						graph.add_edge(bbox1.get_id(), bbox2.get_id())
+
+		return graph
+
+	# axis = 0 (x-axis), 1 (y-axis)
+	def _find_overlap(self, bbox1, bbox2, axis=0):
```

Comment:
Can potentially move this function to `BBox` class to make the code more clear. 
e.g.
```python
class BBox:
    def overlap_with(self, box, axis=0):
        .....
```
No high priority thing, your current code looks not bad.